compiler/packages/babel-plugin-react-compiler/src/Inference/InferReferenceEffects.ts

```diff
@@ -395,24 +395,29 @@ class InferenceState {
         reason,
         context: new Set(),
       });
-      if (value.kind === 'FunctionExpression') {
-        if (
-          this.#env.config.enablePreserveExistingMemoizationGuarantees ||
-          this.#env.config.enableTransitivelyFreezeFunctionExpressions
-        ) {
-          if (value.kind === 'FunctionExpression') {
-            /*
-             * We want to freeze the captured values, not mark the operands
-             * themselves as frozen. There could be mutations that occur
-             * before the freeze we are processing, and it would be invalid
-             * to overwrite those mutations as a freeze.
-             */
-            for (const operand of eachInstructionValueOperand(value)) {
-              const operandValues = this.#variables.get(operand.identifier.id);
-              if (operandValues !== undefined) {
-                this.freezeValues(operandValues, reason);
-              }
+      if (
+        value.kind === 'FunctionExpression' &&
+        (this.#env.config.enablePreserveExistingMemoizationGuarantees ||
+          this.#env.config.enableTransitivelyFreezeFunctionExpressions)
+      ) {
+        for (const operand of value.loweredFunc.func.context) {
+          const operandValues = this.#variables.get(operand.identifier.id);
+          if (operandValues !== undefined) {
+            if (
+              operandValues.size === 1 &&
+              operandValues.values().next().value?.kind === 'DeclareContext'
```

Comment:
hmm why only consider hoisted context variables if there is one operand? could you take the example below and also add some other context reference and break this?

Reply:
Please see `compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/hoisting-setstate.expect.md` for an explanation. I think this check is semantically important, but happy to refactor to make it cleaner

Reply:
TODO: check if we can just "never freeze declare contexts"